src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -71,110 +72,128 @@ final class TrialService
-
-          // TODO: Test the logic below by adding a mock ThurloeDAO whose saveTrialStatus() returns Failure
-          // Save updates to "userInfo's" trial status, authenticating to Thurloe as "managerInfo" user
-          thurloeDao.saveTrialStatus(userInfo.userSubjectId, managerInfo, newStatus) map {
-            case Success(_) =>
-              logger.warn(s"Updated profile saved as $newState; we are done!")
-              StatusUpdate.Success
-            case Failure(ex) =>
-              logger.warn(s"We could not have Thurloe update the new state as $newState; aborting...")
-              StatusUpdate.ServerError(ex.getMessage)
+  private def executeStateTransitions(managerInfo: UserInfo, userEmails: Seq[String],
+                                      statusTransition: (WorkbenchUserInfo, Option[UserTrialStatus]) => UserTrialStatus,
+                                      transitionPostProcessing: (Attempt, Option[UserTrialStatus], UserTrialStatus) => Unit): Future[PerRequestMessage] = {
+    var results: Seq[(String,String)] = Seq()
+    userEmails.foreach { userEmail =>
+     val finalStatus =
+      getUserSubjectId(userEmail) flatMap { subId =>
+        val userInfo = WorkbenchUserInfo(subId, userEmail)
+        thurloeDao.getTrialStatus(subId, managerInfo) flatMap { userTrialStatus =>
+          val newStatus = statusTransition(userInfo, userTrialStatus)
+          if (requiresStateTransition(userTrialStatus, newStatus)) {
+            for {
+              stateResponse <- updateTrialStatus(managerInfo, userInfo, newStatus)
+            } yield {
+              transitionPostProcessing(stateResponse, userTrialStatus, newStatus)
+              StatusUpdate.toName(stateResponse)
+            }
+          } else {
+            logger.warn(
+              s"The user '${userInfo.userEmail}' is already in the trial state of '$newStatus.newState'. " +
+                s"No further action will be taken.")
+            Future.successful(StatusUpdate.toName(StatusUpdate.NoChangeRequired))
           }
         }
       }
-      // TODO: Respond with a more user-friendly error
-      case None => {
-        logger.warn(s"Trial status could not be found for ${userInfo.userEmail}")
+      // Use await here so that multiple Futures don't have a conflict when claiming a billing project
+      val result = (Await.result(finalStatus, scala.concurrent.duration.Duration.Inf))
+      results = results ++ Seq((result, userEmail))
+    }
+    val sorted = results.groupBy(_._1).map { case (k,v) => (k,v.map(_._2))}
+    Future.successful(RequestComplete(sorted))
+  }
 
-        Future(StatusUpdate.Failure)
-      }
+  private def updateTrialStatus(managerInfo: UserInfo,
+                                userInfo: WorkbenchUserInfo,
+                                updatedTrialStatus: UserTrialStatus): Future[StatusUpdate.Attempt] = {
+
+    // TODO: Test the logic below by adding a mock ThurloeDAO whose saveTrialStatus() returns Failure
```

Comment:
Is this for a future story? There are some failure cases handled by `TrialApiServiceSpecThurloeDAO` so this may be handled.